File: problems/2-medium/138-Copy-List-with-Random-Pointer/solution.cpp
#include <unordered_map>
using namespace std;


// Definition for a Node.
class Node {
public:
    int val;
    Node* next;
    Node* random;
    
    Node(int _val) {
        val = _val;
        next = nullptr;
        random = nullptr;
    }
};


Node* copyRandomList(Node* head) {
    unordered_map<Node*, Node*> ori_to_copy = {{nullptr, nullptr}};
    Node* curr_node = head;
    Node* copy_node = nullptr;
    while (curr_node) {
        copy_node = new Node(curr_node->val);
        ori_to_copy.insert({curr_node, copy_node});
        curr_node = curr_node->next;
    }

    curr_node = head;
    while (curr_node) {
        copy_node = ori_to_copy[curr_node];
        copy_node->next = ori_to_copy[curr_node->next];
        copy_node->random = ori_to_copy[curr_node->random];
        curr_node = curr_node->next;
    }   

    return ori_to_copy[head];
}

File: problems/2-medium/138-Copy-List-with-Random-Pointer/solution.py
from typing import Optional

# Definition for a Node.
class Node:
    def __init__(self, x: int, next: 'Node' = None, random: 'Node' = None):
        self.val = int(x)
        self.next = next
        self.random = random


class Solution:
    def copyRandomList(self, head: 'Optional[Node]') -> 'Optional[Node]':
        original_to_copy = {}
        original_to_copy[None] = None

        res = None
        current = head
        while current != None:
            cpy_node = Node(current.val)
            original_to_copy[current] = cpy_node
            current = current.next
        
        current = head
        while current != None:
            res = original_to_copy[current]
            res.next = original_to_copy[current.next]
            res.random = original_to_copy[current.random]
            current=current.next

        return original_to_copy[head]
        


        








        